pkg/api/types.go

```diff
@@ -1508,10 +1512,13 @@ type Toleration struct {
 	// If the operator is Exists, the value should be empty, otherwise just a regular string.
 	Value string `json:"value,omitempty"`
 	// Effect indicates the taint effect to match. Empty means match all taint effects.
-	// When specified, allowed values are NoSchedule and PreferNoSchedule.
+	// When specified, allowed values are NoSchedule, PreferNoSchedule, and NoExecute.
 	Effect TaintEffect `json:"effect,omitempty"`
-	// TODO: For forgiveness (#1574), we'd eventually add at least a grace period
-	// here, and possibly an occurrence threshold and period.
+	// ForgivenessSeconds represents the period of time the toleration (which must be
+	// of effect NoExecute, otherwise this field is ignored) tolerates the taint. By default,
+	// it is not set, which means tolerate the taint forever (do not evict). Zero and
+	// negative values are not allowed.
+	ForgivenessSeconds *int64 `json:"forgivenessSeconds,omitempty"`
```

Comment:
The pattern we've used elsewhere is XXXSeconds. I think ForgivenessSeconds is fine.